95 write a report systems perspective (#105)
Itu-DevOps-2026/ITU-MiniTwit: report/report.md

---

# Introduction

_Authors: _

# System

## Architecture and Design

_Authors: Nikolej_

This project is a forked repository of the Chirp! project [@chirp] developed for the course
"Analysis, Design and Software Architecture (Autumn 2025)".
The MiniTwit application architecture, namely the domain model and codebase structure,

> is inherited unchanged from the Chirp! project.

Author: Feel like this is missing a word.

Reviewer: don't agree

Reviewer 2: I can't find it

Reviewer 3: Could you elaborate? Maybe give a suggestion?

Author: If others says it is fine it is fine

![Illustration of code base](docs/images/onion_architecture.png)
_Illustration of the Chirp! app codebase structure - based on onion architecture (reused from the Chirp! project - not compiled during this course)._

## Deployment

_Authors: Nikolej_

Below is a diagram of the overall deployment architecture of the MiniTwit application.
The system uses two load balancers and two production server instances each running all containerized services and
both read and write to the same DigitalOcean Managed Database.

This setup is designed to prioritize availability.
One can imagine the stakeholders of a social media platform pushing for availability,
since every second of downtime is potential earnings lost.

The dual load balancer (LB) setup was chosen for this reason.
The primary load balancer initially handles all traffic, but is replaced by the secondary in case of failure.
This is enabled through the heartbeat messages exchanged between the two using VRRP.
The secondary then takes ownership of the reserved IP and resumes operation.
Essentially, one LB is active while the other is on standby,

> which keeps the system online and reduces one potential single point of failure.

Reviewer 2: Convince my why the above section should not be in availability and scalling

Reviewer 3: I cannot, for you are probably right. The question is, whether we think it is important enough to move. I personally feel its fine as long as it's included somewhere

Author: Yeah it should probely be moved to that instead

Reviewer 2: Who takes reponsibility for this? 
Be aware that the section below the one I commented on also could possibly be moved to that section

Reviewer 3: I can handle it, but it should probably be done directly on the report branch, once this PR has been merged, since availability and scaling is in the process section

- Grafana - Visualization platform for logs and monitoring
- Prometheus - Monitoring service used to collect and store metrics from infrastructure and applications
- Loki - Log aggregation and storage system
- Alloy - Telemetry collector used to collect and forward logs
- Ubuntu - Linux distribution used on our servers
- MySQL - Relational database management system
- Vagrant - Tool for creating and managing reproducible virtualized development environments

**Static Analysis**

- Hadolint - Linter for Dockerfiles
- CSharpier - C# code formatter
- CodeQL - Static analysis engine used to identify security vulnerabilities and code quality issues
- Roslyn - .NET compiler platform for code analysis
- Docker Scout - Security and vulnerability analysis tool for Docker
- SonarCloud - Cloud-based code quality and security analysis platform
- Codacy - Automated code review and quality monitoring platform

**Development**

- Git - Distributed version control system
- GitHub - Web-based platform for hosting Git repositories
- .NET - Microsoft’s development platform and runtime used to make the minitwit application

## System States

> _Authors: Vitus_

Reviewer: ```suggestion
_Authors: Vitus, Sara_
```

> The project is currently in a good state quality wise.

Reviewer: ```suggestion
The project is currently in a strong state with respect to software quality
```

> We have from the beginning focused on always having our tests and static analysis tools accept incoming changes before being allowed to be merged into a production state. 

Reviewer: ```suggestion
From the outset, we established a workflow in which all incoming changes were required to pass automated tests and static analysis checks before being merged into the production branch. 
```

> Even going as far as to set multiple analysis tools to strict. 

Reviewer: ```suggestion
Several analysis tools were also configured with strict settings to enforce high coding standards.
```

> As of now, the entire project is compliant with these criteria, and can thus be considered stable. 

Reviewer: ```suggestion
At present, the entire project complies with these quality requirements and can thus be considered stable. 
```

> Any crashes we have encountered in this project have been interogated and the leading issues fixed.

Reviewer: ```suggestion
Any crashes or critical issues encountered during development were thoroughly investigated, and the underlying causes were addressed accordingly.
```

There have been a few compromises and temporary bending of these quality gating rules. 
The project is developed on top of the Chirp! platform, which was specifically made for dotnet 8, that has become deprecated and outdated since then, having moved onto dotnet 10.
This means that the Docker Scout scan of the image strongly recommends us to update. We have therefore chosen to make the analysis ignore said category explicitly. 

> This is not a breaking issue, as long as we keep this in mind.

Reviewer: ```suggestion
There have been a few necessary compromises to these quality gating rules. 
The project is developed on top of the Chirp! platform, which was originally designed for .NET 8. Since then, newer .NET versions have been released, and .NET 8 is not considered outdated.
As a result, As a result, the Docker Scout analysis consistently recommends upgrading the runtime environment.
Given the project constraints and dependencies on the underlying platform, we chose to explicitly exclude this specific category from the analysis checks. While this represents a deviation from the otherwise strict quality requirements, it is not considered a critical issue as long as the limitation is acknowledged and taken into account during maintenance and future development.
```
